nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,222 @@
-                               alice_signature: Signature = None,  # TODO: Why is this unused?
-                               session=None
-                               ) -> PolicyArrangement:
-        """
-        Creates a PolicyArrangement to the Keystore.
-
-        :return: The newly added PolicyArrangement object
-        """
-        session = session or self._session_on_init_thread
-
-        alice_key_instance = session.query(Key).filter_by(key_data=bytes(alice_verifying_key)).first()
-        if not alice_key_instance:
-            alice_key_instance = Key.from_umbral_key(alice_verifying_key, is_signing=True)
-
-        new_policy_arrangement = PolicyArrangement(
-            expiration=expiration,
-            id=arrangement_id,
-            kfrag=kfrag,
-            alice_verifying_key=alice_key_instance,
-            alice_signature=None,
-            # bob_verifying_key.id  # TODO: Is this needed?
-        )
-
-        session.add(new_policy_arrangement)
-        self.__commit(session=session)
-        return new_policy_arrangement
-
-    def get_policy_arrangement(self, arrangement_id: bytes, session=None) -> PolicyArrangement:
-        """
-        Retrieves a PolicyArrangement by its HRAC.
-
-        :return: The PolicyArrangement object
-        """
-        session = session or self._session_on_init_thread
-        policy_arrangement = session.query(PolicyArrangement).filter_by(id=arrangement_id).first()
-        if not policy_arrangement:
-            raise NotFound("No PolicyArrangement {} found.".format(arrangement_id))
-        return policy_arrangement
-
-    def get_all_policy_arrangements(self, session=None) -> List[PolicyArrangement]:
-        """
-        Returns all the PolicyArrangements
-
-        :return: The list of PolicyArrangement objects
-        """
-        session = session or self._session_on_init_thread
-        arrangements = session.query(PolicyArrangement).all()
-        return arrangements
-
-    def attach_kfrag_to_saved_arrangement(self, alice, id_as_hex, kfrag, session=None):
-        session = session or self._session_on_init_thread
-        policy_arrangement = session.query(PolicyArrangement).filter_by(id=id_as_hex.encode()).first()
+    pass
 
-        if policy_arrangement is None:
-            raise NotFound("Can't attach a kfrag to non-existent Arrangement {}".format(id_as_hex))
 
-        if policy_arrangement.alice_verifying_key.key_data != alice.stamp:
-            raise alice.SuspiciousActivity
+class DatastoreKey(NamedTuple):
+    """
+    Used for managing keys when querying the datastore.
+    """
+    record_type: Optional[str] = None
+    record_field: Optional[str] = None
+    record_id: Optional[Union[bytes, int]] = None
 
-        policy_arrangement.kfrag = bytes(kfrag)
-        self.__commit(session=session)
+    @classmethod
+    def from_bytestring(cls, key_bytestring: bytes) -> 'DatastoreKey':
+        key_parts = key_bytestring.decode().split(':')
```

Comment:
You can't have `:` in a class name or an attribute, so that's not possible there. It's only possible in the ID. There's currently no checks in `DatastoreRecord` for invalid characters in the record ID, so this would make it such that any call to `query_by` that hits this ID will error resulting in an impossible to finish query. Good catch.

I don't think the solution is to prevent `:` from being in the ID (that may actually be useful at some point for encoding further information in the key string). Additionally, I much prefer to avoid writing branching logic for things like bad characters.

What do you think about simply putting the return of this method in a `try/except`, and when we get a `TypeError` for passing too many pieces we simply provide the first three? This allows us to ensure that all key strings are evaluated equitably, and additionally lets us avoid having to concern ourselves with the minutiae of what a developer names their IDs all while keeping an extremely flexible key structure.

Reply:
That would mean that only the first part of such hypothetical record ID is considered. That's fine as long as it's documented somewhere. I also had an alternative idea: what if you simply use ByteStringSplitter here?